lib/concurrent-ruby/concurrent/collection/ruby_timeout_queue.rb

```diff
@@ -0,0 +1,57 @@
+module Concurrent
+  module Collection
+    # @!visibility private
+    # @!macro ruby_timeout_queue
+    class RubyTimeoutQueue < ::Queue
+      def initialize(*args)
+        if RUBY_VERSION >= '3.2'
+          raise "#{self.class.name} is not needed on Ruby 3.2 or later, use ::Queue instead"
+        end
+
+        super(*args)
+
+        @mutex = Mutex.new
+        @cond_var = ConditionVariable.new
+      end
+
+      def push(obj)
+        @mutex.synchronize do
+          super(obj)
+          @cond_var.signal
+        end
+      end
+      alias_method :enq, :push
+      alias_method :<<, :push
+
+      def pop(non_block = false, timeout: nil)
+        if non_block && timeout
+          raise ArgumentError, "can't set a timeout if non_block is enabled"
+        end
+
+        if non_block
+          super(true)
+        elsif timeout
+          @mutex.synchronize do
+            deadline = Concurrent.monotonic_time + timeout
+            while (now = Concurrent.monotonic_time) < deadline # handle spurious wakeups
+              begin
+                return super(true)
+              rescue ThreadError
+                # empty, wait until woken up
+              end
+
+              @cond_var.wait(@mutex, deadline - now)
+            end
+
+            nil
```

Comment:
```suggestion
            while (now = Concurrent.monotonic_time) < deadline and empty?
              @cond_var.wait(@mutex, deadline - now)
            end
            begin
              return super(true)
            rescue ThreadError
              # still empty
              nil
            end
```
This way we avoid many `ThreadError` exceptions, and we're under the mutex so there is no race possible between the `empty?` and the `pop`/super.

Also importantly even if the timeout is very small (or even 0) it still tries to pop once.

Reply:
[Addressed](https://github.com/ruby-concurrency/concurrent-ruby/compare/9739e9ad72a9e9f7789d911d50dbe25ede887d29..6aa4ee5bea7bec58d3186c404533c7b86e729894).